### PR TITLE
feat(core): compound ast-grep schema + deterministic manifest hashing (#1407)

### LINKED ISSUE #1407: feat(core): extend CompiledRule schema with astGrepYamlRule field
## Context

Proposal 226 requires `CompiledRule` to hold compound ast-grep rules alongside the existing flat `astGrepPattern` string. This ticket lands the schema change and extends compile-time validation.

## Dependencies

Blocked by the `@ast-grep/napi` YAML validation spike (first ticket in this milestone). We need the spike's findings to confirm the exact shape of YAML rule objects before locking the schema.

## Scope

1. Add `astGrepYamlRule?: AstGrepYamlRule` to `CompiledRule` (where `AstGrepYamlRule` is a Zod schema covering the YAML rule object shape validated by the spike).
2. Preserve `astGrepPattern` as-is. The two are mutually exclusive per rule; validator rejects rules that set both or neither for ast-grep-mode rules.
3. Extend `validateAstGrepPattern` (introduced in #1349) to accept and validate YAML rule objects via `@ast-grep/napi`.
4. Ensure compile manifest hash input includes `astGrepYamlRule` content so manifest drift detection (#1337 / #1348) still works for YAML rules.

## Out of Scope

- Engine runtime execution of compound rules (separate ticket).
- Compiler prompt tuning (separate ticket).
- Bulk recompile of archived rules tagged `upgradeTarget: compound` (phase 4 on the epic).

## Acceptance Criteria

- [ ] `CompiledRule` schema compiles with strict TypeScript; new field documented in schema comments.
- [ ] Validator rejects rules with both `astGrepPattern` and `astGrepYamlRule` set with a fail-loud `TotemError`.
- [ ] Validator rejects syntactically invalid YAML rules via the spike-proven path.
- [ ] Compile manifest hash input includes `astGrepYamlRule` when present.
- [ ] Unit tests cover both the dual-field rejection and the YAML validation path.

## References

- Epic: mmnto-ai/totem-strategy#81
- Proposal: `.strategy/proposals/active/226-compound-ast-grep-rules.md`
- Prior validation work: #1349

### PR DESCRIPTION
**Pull request by satur8d**
## Mechanical Root Cause

ADR-087 locked the architecture for compound ast-grep rule support. This PR lands the schema half: split `astGrepPattern` into a flat-string field and a new structural `astGrepYamlRule: NapiConfig` field with Zod-level mutual exclusion, route the new field through the existing `validateAstGrepPattern`, and add deterministic manifest hashing so two compile runs that emit the same compound rule with different key orders cannot trip `verify-manifest`. `badExample` fields are added to both compiler-output and compiled-rule schemas as optional for 1.14.9 (gate wired in #1408).

Design doc at `.totem/specs/1407.md` was approved before code was written. All three open questions resolved on (a), (b), (b).

## Fix Applied

Five bisectable commits on this branch:

1. **`acd6dd39`** — `feat(core): add NapiConfigSchema and schema-level mutual exclusion for ast-grep rules` — new `NapiConfigSchema`, `AstGrepYamlRuleSchema` alias, new `astGrepYamlRule` + optional `badExample` fields on both `CompiledRuleSchema` and `CompilerOutputSchema`, shared `superRefine` gating mutual exclusion and required-one-of. 16 new invariants in `compiler-schema.test.ts`.
2. **`66a06b89`** — `feat(core): route astGrepYamlRule through validateAstGrepPattern` — `buildCompiledRule` picks whichever ast-grep field the LLM emits, hands it to `validateAstGrepPattern`; `engineFields` routes strings to `astGrepPattern` and objects to `astGrepYamlRule`. `isSelfSuppressing` stringify path preserved and tested against deep-leaf `totem-ignore` per the design doc invariant.
3. **`baf02bc0`** — `feat(core): canonical JSON stringify for deterministic manifest hashing` — new pure `canonicalStringify` utility sorts keys at every depth; `generateOutputHash` only switches to canonical serialisation when the rules file contains `astGrepYamlRule`, so every existing user's `compile-manifest.json` stays valid after upgrading without a forced recompile.
4. **`54604c65`** — `test(cli): migrate run-compiled-rules fixtures to split ast-grep fields` — two TOTEM_LITE tests had to move their match source from `pattern` into `astGrepPattern` after the superRefine went live.
5. **`4dc05b25`** — `fix(core): canonicalStringify throws on bare undefined` — addresses the GCA INFO finding pre-emptively. Per Tenet 4 (Fail Loud, Never Drift): undefined is not a JSON value, so the contract throws rather than returns `undefined` as a string.

## Out of Scope

- **Engine runtime routing for compound rules.** `applyAstRulesToAdditions` at `rule-engine.ts:415` still filters on `r.astGrepPattern` only. A compound rule would never reach `findAll()` at runtime today. Scoped to #1408 per the design doc. One `rule-engine.test.ts` fail-closed test had to swap its compound fixture for a runtime-invalid string pattern because the compound path is unreachable; #1408 will restore the compound fixture.
- **Smoke-test runner for `badExample`.** The field is stored but not executed. Gate wired in #1408.
- **Compiler prompt changes.** Sonnet is not taught to emit `astGrepYamlRule` or `badExample` here. Lives in #1409.
- **`extractManualPattern` `badExample` extraction.** Pipeline 1 stays untouched per locked question 2.
- **Object-walker for `isSelfSuppressing`.** Design doc pinned the existing stringify path as correct; regression test locks that guarantee so a future walker refactor cannot regress it.
- **Auto-archive on napi upgrade.** Break-loud contract per #1345 stays intact. Locked question 3.

## Tests Added/Updated

- [x] 25 new tests (core: 1059 → 1084)
- [x] 16 invariants on schema mutual-exclusivity locked in `compiler-schema.test.ts`
- [x] `isSelfSuppressing` stringify path test for compound rules with deep `totem-ignore` leaves
- [x] `canonicalStringify` stability across 6 nesting patterns plus primitive handling plus `undefined` throw
- [x] `generateOutputHash` determinism across scrambled key orders for compound rules
- [x] Existing test migrations in `run-compiled-rules.test.ts` (fixture-only, no semantic change)

**Verification:**

- `pnpm -r test`: all 2819 green (was 2794)
- `pnpm exec totem lint`: PASS (21 warnings, all flat-rule false positives in comments/test strings, 0 errors)
- `pnpm exec totem review`: PASS (2 INFO findings, non-blocking — both are comment-accuracy nits on the guard logic which is itself correct)
- `pnpm exec totem verify-manifest`: PASS (410 rules, hashes match — backward compat path worked)
- `tsc --noEmit`: clean across all packages

## Related Tickets

Closes #1407

Epic: mmnto-ai/totem-strategy#81
ADR: `.strategy/adr/adr-087-compound-ast-grep-rules.md`
Spike findings: `packages/core/spikes/compound-ast-grep/findings.md` (merged in #1410)
Design doc: `.totem/specs/1407.md`

Blocks: #1408 (engine runtime + smoke gate + Pipeline 1 extractor). Unblocks phase 4 batch-recompile of the `upgradeTarget: compound` archived queue after #1408 and #1409 ship.

### CODE REVIEW

File: packages/core/src/compiler-schema.ts [hunk 1]
```diff
@@ -15,8 +50,27 @@ export const CompiledRuleSchema = z.object({
...
   /** Tree-sitter S-expression query (required when engine is 'ast') */
   astQuery: z.string().optional(),
-  /** ast-grep pattern — string for simple patterns, object for compound rules (has/inside/not) */
-  astGrepPattern: z.union([z.string(), z.record(z.unknown())]).optional(),
+  /**
+   * Flat ast-grep pattern source (a single JS/TS expression). Mutually
+   * exclusive with `astGrepYamlRule` when `engine === 'ast-grep'`; the
+   * superRefine below enforces that.
+   */
+  astGrepPattern: z.string().optional(),
```
[the remark this answers is not in the document]
Comment (satur8d) [reply packages/core/src/compiler-schema.ts:58]: @gemini-code-assist Dismissed. The premise does not hold: no compiled rule in any user's on-disk `compiled-rules.json` uses the compound object form for `astGrepPattern` today. The compiler has never emitted compound rules — that capability is the entire purpose of this PR. The pre-#1407 union type `z.union([z.string(), z.record(z.unknown())])` was speculative infrastructure that never saw production emission, which is exactly why the spike #1406 had to empirically prove the runtime would accept the shape.

Verified: `git grep astGrepYamlRule -- ':(glob)**/compiled-rules.json'` and `git grep 'astGrepPattern.*\{' -- ':(glob)**/compiled-rules.json'` both return zero matches across the monorepo. Zero migration risk. Loader-level compatibility shim would be dead code on arrival.

File: packages/core/src/compile-manifest.ts [hunk 2]
```diff
@@ -75,15 +75,92 @@ export function generateInputHash(lessonsDir: string): string {
...
+        const parsed: unknown = JSON.parse(raw);
+        if (typeof parsed === 'object' && parsed !== null) {
+          payload = canonicalStringify(parsed);
+        }
+      } catch {
+        // Malformed JSON: keep the raw byte stream so verify-manifest
+        // still surfaces the corruption via a mismatch rather than
+        // crashing here.
+      }
+    }
```
[the remark this answers is not in the document]
Comment (satur8d) [reply packages/core/src/compile-manifest.ts:190]: @gemini-code-assist Valid finding, fixed in commit e4ec7ed3. `generateOutputHash` now parses the JSON and checks the actual `astGrepYamlRule` field presence on each rule via a new local `hasCompoundRule` helper, rather than substring-matching the raw bytes. The substring check remains as a cheap O(1) early-out before the O(N) parse, so the happy path is unchanged for manifests that do not mention the field at all.

Regression test added that writes a manifest containing the literal bytes `"astGrepYamlRule"` inside a lesson message and asserts the raw-byte-stream path is preserved (hash equals sha256 of the file contents). Pre-#1407 CLIs' stored hashes stay valid.

File: packages/core/src/compile-manifest.ts [hunk 3]
```diff
@@ -75,15 +75,92 @@ export function generateInputHash(lessonsDir: string): string {
...
+  }
+  const record = value as Record<string, unknown>;
+  const keys = Object.keys(record)
+    .filter((k) => record[k] !== undefined)
+    .sort();
+  const parts: string[] = [];
+  for (const k of keys) {
+    parts.push(JSON.stringify(k) + ':' + canonicalStringify(record[k]));
+  }
+  return '{' + parts.join(',') + '}';
```
[the remark this answers is not in the document]
Comment (satur8d) [reply packages/core/src/compile-manifest.ts]: Fixed via JSDoc tightening in commit e4ec7ed3. The contract comment now explicitly names the plain-JSON-only boundary (`JSON.parse()` output: `null`, booleans, numbers, strings, arrays, plain object literals), lists the specific classes of divergent inputs (`Date` serialises to `{}` rather than ISO string, `[undefined]` throws rather than yielding `[null]`), and warns future callers away from using this as a drop-in `JSON.stringify` replacement. Bare `undefined` throws `TotemParseError` per the earlier fix in `4dc05b25`. Runtime validation for class instances would reject legitimate test fixtures that lint us into `Object.create(null)` corner cases and was judged disproportionate for a caller set that is entirely `JSON.parse`-bounded.

File: packages/core/src/compiler-schema.ts [hunk 4]
```diff
@@ -1,8 +1,43 @@
...
+ * `any`, `not`, `inside`, `has`, `precedes`, `follows`). Rather than
+ * modelling the full recursive schema with `z.lazy()`, we accept any
+ * object shape here and lean on napi to reject malformed trees. That
+ * keeps the Zod layer cheap and the authoritative check centralized.
+ */
+export const NapiConfigSchema = z
+  .object({
+    rule: z.record(z.unknown()),
+  })
+  .passthrough();
```
[the remark this answers is not in the document]
Comment (satur8d) [reply packages/core/src/compiler-schema.ts:24]: Dismissed. We are pinned at `zod@^3.24.0` across all three packages (verified in `packages/core/package.json:37`, `packages/cli/package.json:35`, `packages/mcp/package.json:29`). On Zod 3.24, `z.record(z.unknown())` is the correct one-argument form. The two-argument signature is a Zod 4 breaking change we have not adopted.

The 2819-test suite passes green including the 16 new schema invariants on this file, which would have failed loudly if the Zod parse were broken.

File: packages/core/src/compiler.ts [hunk 5]
```diff
@@ -226,12 +231,24 @@ function normalizeShallowGlob(glob: string): string {
...
+      // Route strings to the flat field and objects to the compound field.
+      // mmnto/totem#1407 split the fields for explicit mutual exclusion; the
+      // caller is responsible for passing exactly one shape. The superRefine
+      // on CompiledRuleSchema gates the persisted rule.
+      if (typeof pattern === 'string') {
+        return { pattern: '', astGrepPattern: pattern };
+      }
+      return { pattern: '', astGrepYamlRule: pattern as AstGrepYamlRule };
     case 'ast':
       return { pattern: '', astQuery: String(pattern) };
```
[the remark this answers is not in the document]
Comment (satur8d) [reply packages/core/src/compiler.ts:284]: Valid finding, fixed in commit e4ec7ed3. Added three TypeScript overloads: `'regex' | 'ast'` accepts string only, `'ast-grep'` accepts `string | Record<string, unknown>`, and a wildcard overload for callers whose `engine` discriminator is resolved at runtime (`buildManualRule` in `compile-lesson.ts` forwards the union literal). The wildcard mirrors the implementation signature. Implementation body unchanged; this is a pure type-safety tightening — the `superRefine` on `CompiledRuleSchema` remains the load-bearing gate.
